platform/engineering/design-docs/2020-08-11-redirects-from-cms.md

```diff
@@ -0,0 +1,224 @@
+
+The current state of managing redirects for content, involves content editors and a developer coordinating and choreographing the release of the contnet with the addition of the redirect(s).  It is time intensive and risk prone.  VA needs a redirect system that is easier to manage and less reliant on developers timing releases perfectly to match.
+
+### Guiding Principles
+1. Veterans should not get 404 errors.
+2. Content on VA.gov should not have broken links to content also on VA.gov.
+3. As the CMS scales from ~20 editors and daily content releases, up to ~500 editors and continuous content releases, we can not allow a broken link to prevent content releases.
+4. Developers currently waste far too much time chasing down broken links.
+
+
+### High Level Design
+
+Three parts of the system work together:
+
+#### I. CMS
+At the highest level, the CMS manages redirects already.  It tracks them, sorts out and prevents circular references (infinite redirects), creates new redirects for moved pages and can handle query parameters, if needed for legacy handling. The CMS even provides a nice user interface for adding, deleting, and editing source, destination and status codes.  The CMS currently makes redirects available to both GraphQL and the static file content exports.  Only one minor change is needed on the CMS side to translate node/# paths to the proper alias.  We simply need a way for the Front-End to act on the redirects in the CMS. The CMS treats redirects like content, and the Front-End should too.  When a title changes, in most cases, the URL changes, so a redirect is needed.  Any Front-End system that does not account for that is a 404 factory, which is our current condition.
+
+#### II. Vets-website content creation.
+Local environments that are not AWS based would need to function correctly with automated tests and human users.  So one part of this strategy is to create client side redirects by creating an html file with a meta refresh tag
+`<meta http-equiv="refresh" content="0; URL='/new/path/destination'" />` for each redirect object found in the content export.  This will happen at the same time that content html pages are built.
+
+#### III. Vets-website deploy to AWS
+During the vets-website deploy to AWS S3, the system currently generates an object for each static page of the VA.gov website.  Via the same object creation process we can add an [S3 object with redirect metadata](https://docs.aws.amazon.com/AmazonS3/latest/dev/how-to-page-redirect.html) that describes both the destination and the return code for each HTML page that contains a meta-refresh tag. These S3 objects get discarded and rebuilt on each content release, so they would **always** be in sync with the content from the CMS.  By adding these, the website when running in an AWS environment would have the more desireable server side redirects instead of the browser-based `meta-refresh` redirects.
```

Comment:
does this match how we serve content from S3? could we actually use this redirect functionality on S3, and would that work with our revproxy setup?

cc @cvalarida @billfienberg while Rian's out

Reply:
Right now, we're using `aws s3 sync` to sync up the contents of the build directory in Jenkins to the S3 bucket. I don't know offhand how we'd manage the creation, updating, and removal of these redirects, but it seems feasible, given the documentation Steve links to.

This definitely needs an experiment run to verify we can do what we'd need to.

Reply:
They would not need to be removed, as both html pages and these redirects would all be created anew on each release.

Reply:
[HTML redirects](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#Alternative_way_of_specifying_redirections) should work with the current nginx setup. I don't think headers pass through but this would be inside of the body which nginx only changes to support CSP. Either way we could do a simple test on staging by adding a document with this metatag.  

It's preferable to use an HTTP redirect but I don't think there's a tangible difference. Users will get a 200 instead of a 301 but the browser will treat it like a permanent redirect. 

This approach doesn't couple us to S3 since it's just creating an html document with a metatag in it.

Reply:
@rianfowler returning misleading headers is not great for SEO. It would also make it impossible for our own link checkers (both on front end and inside the CMS) to surface whether something is pointing to a redirected page, it would CURL the page to see if it exists, it would get a misleading 200 response.  
We spent a lot of time building the functionality to detect redirects and warn editors
![image](https://user-images.githubusercontent.com/5752113/91895665-a1e1c000-ec65-11ea-9ef7-57644974232d.png)
This shortcut would prevent that from working.

Reply:
A 200 response will impact SEO as the browser and search engine think all is well, but we need search engines to know that a page has been permanently redirected so that equity is passed on and SERPs are updated.   We need 301s to retain the SEO value of pages.